nexus/src/app/disk.rs

```diff
@@ -181,15 +208,52 @@ impl super::Nexus {
-        if params.size.to_bytes() > MAX_DISK_SIZE_BYTES {
-            return Err(Error::invalid_value(
-                "size",
-                format!(
-                    "total size must be less than {}",
-                    ByteCount::try_from(MAX_DISK_SIZE_BYTES).unwrap()
-                ),
-            ));
+        // Check for disk type specific restrictions
+        match &params {
+            params::DiskCreate::Crucible { .. } => {
+                // Reject disks where the size is greated than
+                // MAX_DISK_SIZE_BYTES. This restriction will be changed or
+                // removed when multi-subvolume Volumes can be created by Nexus,
+                // or if the region allocation algorithm changes.
+                if params.size().to_bytes() > MAX_DISK_SIZE_BYTES {
+                    return Err(Error::invalid_value(
+                        "size",
+                        format!(
+                            "total size must be less than {}",
+                            ByteCount::try_from(MAX_DISK_SIZE_BYTES).unwrap()
+                        ),
+                    ));
+                }
+            }
+
+            params::DiskCreate::LocalStorage { .. } => {
+                // If a user requests some outlandish number of TB for local
+                // storage, and there isn't a sled allocation that can fulfill
+                // this, instance create will work but instance start (which
+                // performs the actual vmm + local storage dataset allocation)
+                // will fail.
+                //
+                // There's an opportunity to consult the latest inventory
+                // collection here, and if there aren't zpools that are as large
+                // as this local storage request, reject it. If the user
+                // mistakenly asks for 30 TB instead of 3 TB, such a check would
+                // prevent the instance start request from succeeding. Later, if
+                // new disks are inserted that have a much larger capacity, then
+                // the same instance create request that could use that larger
+                // capacity.
+                //
+                // We don't want to reject creating a bunch of instances
+                // requesting local storage, where if all of them were started
+                // it wouldn't work, but the user's intention is not to ever
+                // start them all at the same time. But we also don't reserve or
+                // allocate any local storage here, so we can't prevent users
+                // from requesting too much local storage to ever start the
+                // instances for.
+                //
+                // TODO consult the latest inventory collection if it isn't too
+                // expensive, and reject local storage disk requests that are
+                // too large to be served by any sled.
```

Comment:
i presume we intend to do that in a subsequent branch, rather than as part of this PR?

Reply:
This is still a TODO I haven't written anything for yet.